sdks/python/apache_beam/transforms/ptransform.py

```diff
@@ -97,29 +100,96 @@ def visit(self, node, replacements):
       return self.visit_nested(node, replacements)
 
 
+# Caches to allow for materialization of values when executing a pipeline
+# in-process, in eager mode.  This cache allows the same _MaterializedResult
+# object to be accessed and used despite Runner API round-trip serialization.
+_pipeline_materialization_cache = collections.defaultdict(dict)
+_pipeline_materialization_lock = threading.Lock()
+
+
+def _allocate_materialized_result(pipeline):
+  with _pipeline_materialization_lock:
+    pipeline_id = id(pipeline)
+    result_id = len(_pipeline_materialization_cache[pipeline_id])
+    result = _MaterializedResult(pipeline_id, result_id)
+    _pipeline_materialization_cache[pipeline_id][result_id] = result
+    return result
+
+
+def _get_materialized_result(pipeline_id, result_id):
+  with _pipeline_materialization_lock:
+    return _pipeline_materialization_cache[pipeline_id][result_id]
+
+
+def _release_materialized_results(pipeline):
+  with _pipeline_materialization_lock:
+    pipeline_id = id(pipeline)
+    if pipeline_id in _pipeline_materialization_cache:
+      del _pipeline_materialization_cache[pipeline_id]
+
+
+class _MaterializedResult(object):
+  def __init__(self, pipeline_id, result_id):
+    self._pipeline_id = pipeline_id
+    self._result_id = result_id
+    self.elements = []
+
+  def __reduce__(self):
+    # When unpickled (during Runner API roundtrip serailization), get the
+    # _MaterializedResult object from the cache so that values are written
+    # to the original _MaterializedResult when run in eager mode.
+    return (_get_materialized_result, (self._pipeline_id, self._result_id))
+
+
 class _MaterializedDoOutputsTuple(pvalue.DoOutputsTuple):
-  def __init__(self, deferred, pvalue_cache):
+  def __init__(self, deferred, results_by_tag):
     super(_MaterializedDoOutputsTuple, self).__init__(
         None, None, deferred._tags, deferred._main_tag)
     self._deferred = deferred
-    self._pvalue_cache = pvalue_cache
+    self._results_by_tag = results_by_tag
 
   def __getitem__(self, tag):
-    # Simply accessing the value should not use it up.
-    return self._pvalue_cache.get_unwindowed_pvalue(
-        self._deferred[tag], decref=False)
+    return self._results_by_tag[tag].elements
```

Comment:
<!--thread_id:cc_164253415_t; commit:41f28e254ebcb4931fc08784830f04d4d025b1a5; resolved:1-->
<!--section:context-quote-->
> **robertwb** wrote:
> Should it be an error rather than an empty value to get an unknown tag?

<!--section:body-->
Thanks, I've made this more explicit and added a test.